File: lib/data/exceptions/calil_api_exception.dart
abstract class CalilApiException implements Exception {
  final String message;
  const CalilApiException(this.message);

  @override
  String toString() => '$runtimeType: $message';
}

class CalilNetworkException extends CalilApiException {
  const CalilNetworkException(super.message);
}

class CalilHttpException extends CalilApiException {
  final int statusCode;
  const CalilHttpException(super.message, {required this.statusCode});
}

class CalilParseException extends CalilApiException {
  const CalilParseException(super.message);
}

class CalilTimeoutException extends CalilApiException {
  const CalilTimeoutException(super.message);
}

File: src/data/exceptions/calilApiException.ts
/**
 * Base exception for all Calil API related errors.
 * Mirrors the Dart `CalilApiException` hierarchy.
 */
export class CalilApiException extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CalilApiException';
    // Restore prototype chain for reliable `instanceof` checks after transpile.
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class CalilNetworkException extends CalilApiException {
  constructor(message: string) {
    super(message);
    this.name = 'CalilNetworkException';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class CalilHttpException extends CalilApiException {
  readonly statusCode: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = 'CalilHttpException';
    this.statusCode = statusCode;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class CalilParseException extends CalilApiException {
  constructor(message: string) {
    super(message);
    this.name = 'CalilParseException';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class CalilTimeoutException extends CalilApiException {
  constructor(message: string) {
    super(message);
    this.name = 'CalilTimeoutException';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
